pkg/store/bucket.go

```diff
@@ -2928,6 +2930,12 @@ func (r *bucketIndexReader) fetchPostings(ctx context.Context, keys []labels.Lab
 		// Fetch from object storage concurrently and update stats and posting list.
 		g.Go(func() error {
 			begin := time.Now()
+			stats := new(queryStats)
+			defer func() {
+				r.mtx.Lock()
+				r.stats = r.stats.merge(stats)
```

Comment:
Maybe we could just move this usage of `r.mtx` inside of `r.stats.merge` i.e. `r.stats.merge()` could use a separate mutex underneath? Then we would avoid having to do this dance the whole time.

Reply:
```
func (s queryStats) merge(o *queryStats) *queryStats
```

This is the current `merge` function and it takes the value as receiver (not pointer). This seems not that easy to add a lock to `queryStats` struct itself. WDYT?
Is it fine to keep the current code

Reply:
Why not convert it into `func (s *queryStats) merge(o *queryStats) *queryStats`? :thinking:

Reply:
I thought the idea is to return a new queryStats object every time. In this case, we will return the same one. But I guess it should be ok, let me try.